From 92469c57a004d920ce4207e4743c1a6882bcd1c5 Mon Sep 17 00:00:00 2001
From: Herbert Xu <herbert@gondor.apana.org.au>
Date: Mon, 1 Sep 2014 13:10:39 -0400
Subject: [crypto] drbg: fix failure of generating multiple of 2**16 bytes

Message-id: <E1XORNX-00089V-Pu@gondolin.me.apana.org.au>
Patchwork-id: 90076
O-Subject: [PATCH 18/19] crypto: drbg - fix failure of generating multiple of 2**16 bytes
Bugzilla: 1011562
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: Rafael Aquini <aquini@redhat.com>

RHEL7 Bugzilla #1011562

Upstream commit: ce5481d01f67ad304908ec2113515440c0fa86eb

The function drbg_generate_long slices the request into 2**16 byte
or smaller chunks. However, the loop, however invokes the random number
generation function with zero bytes when the request size is a multiple
of 2**16 bytes. The fix prevents zero bytes requests.

Signed-off-by: Stephan Mueller <smueller@chronox.de>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/crypto/drbg.c b/crypto/drbg.c
index 8527eec..9c50c2d 100644
--- a/crypto/drbg.c
+++ b/crypto/drbg.c
@@ -1500,7 +1500,7 @@ static int drbg_generate_long(struct drbg_state *drbg,
   if (0 >= tmplen)
    return tmplen;
   len += tmplen;
- } while (slice > 0);
+ } while (slice > 0 && (len < buflen));
  return len;
 }
 
-- 
1.7.1